File: iec61131lang/src/main/java/edu/kit/iti/formal/automation/analysis/FindDataTypes.java
package edu.kit.iti.formal.automation.analysis;

/*-
 * #%L
 * iec61131lang
 * %%
 * Copyright (C) 2017 Alexander Weigl
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import edu.kit.iti.formal.automation.scope.Scope;
import edu.kit.iti.formal.automation.st.ast.*;
import edu.kit.iti.formal.automation.st.util.AstVisitor;
import lombok.RequiredArgsConstructor;

/**
 * @author Alexander Weigl, Augusto Modanese
 * @version 1 (25.06.17)
 */
@RequiredArgsConstructor
public class FindDataTypes extends AstVisitor {
    private final Scope globalScope;

    @Override
    public Object visit(ProgramDeclaration programDeclaration) {
        setParentScope(programDeclaration);
        globalScope.registerProgram(programDeclaration);
        programDeclaration.getActions().forEach((n,a)->programDeclaration.getScope().registerAction(a));
        return null;
    }

    @Override
    public Object visit(FunctionBlockDeclaration functionBlockDeclaration) {
        setParentScope(functionBlockDeclaration);
        globalScope.registerFunctionBlock(functionBlockDeclaration);
        functionBlockDeclaration.getActions().forEach((n, a) -> functionBlockDeclaration.getScope().registerAction(a));
        return null;
    }

    @Override
    public Object visit(FunctionDeclaration functionDeclaration) {
        setParentScope(functionDeclaration);
        globalScope.registerFunction(functionDeclaration);
        return null;
    }

    @Override
    public Object visit(SubRangeTypeDeclaration subRangeTypeDeclaration) {
        globalScope.registerType(subRangeTypeDeclaration);
        return null;
    }

    @Override
    public Object visit(VariableDeclaration variableDeclaration) {
        //weigl: do not register anonymous datatype, or references to data types.
        /*
        if (variableDeclaration.getTypeDeclaration() instanceof ArrayTypeDeclaration)
            variableDeclaration.getTypeDeclaration().accept(this);
        return super.visit(variableDeclaration);
        */
        return null;
    }

    @Override
    public Object visit(SimpleTypeDeclaration simpleTypeDeclaration) {
        globalScope.registerType(simpleTypeDeclaration);
        return super.visit(simpleTypeDeclaration);
    }

    @Override
    public Object visit(PointerTypeDeclaration ptd) {
        globalScope.registerType(ptd);
        return super.visit(ptd);
    }

    @Override
    public Object visit(ClassDeclaration clazz) {
        setParentScope(clazz);
        globalScope.registerClass(clazz);
        return super.visit(clazz);
    }

    @Override
    public Object visit(InterfaceDeclaration interfaceDeclaration) {
        setParentScope(interfaceDeclaration);
        globalScope.registerInterface(interfaceDeclaration);
        return super.visit(interfaceDeclaration);
    }

    @Override
    public Object visit(ArrayTypeDeclaration arrayTypeDeclaration) {
        globalScope.registerType(arrayTypeDeclaration);
        return null;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Object visit(EnumerationTypeDeclaration enumerationTypeDeclaration) {
        globalScope.registerType(enumerationTypeDeclaration);
        return null;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Object visit(StringTypeDeclaration stringTypeDeclaration) {
        globalScope.registerType(stringTypeDeclaration);
        return null;
    }

    @Override
    public Object visit(TypeDeclarations typeDeclarations) {
        for (TypeDeclaration td : typeDeclarations)
            td.accept(this);
        return null;
    }

    public Object visit(StructureTypeDeclaration structureTypeDeclaration) {
        globalScope.registerType(structureTypeDeclaration);
        return null;
    }

    @Override
    public Object visit(GlobalVariableListDeclaration gvl) {
        globalScope.addVariables(gvl.getScope());
        return super.visit(gvl);
    }

    private void setParentScope(TopLevelScopeElement tle) {
        tle.getScope().setParent(globalScope);
    }
}

File: iec61131lang/src/main/kotlin/edu/kit/iti/formal/automation/analysis/FindDataTypes.kt
package edu.kit.iti.formal.automation.analysis

import edu.kit.iti.formal.automation.scope.Scope
import edu.kit.iti.formal.automation.st.ast.*
import edu.kit.iti.formal.automation.st.util.AstVisitorWithScope

/**
 * @author Alexander Weigl, Augusto Modanese
 * @version 1 (25.06.17)
 */
class FindDataTypes(val globalScope: Scope) : AstVisitorWithScope<Any>() {
    override fun visit(pd: ProgramDeclaration): Any? {
        goIntoScope(pd)
        scope.registerProgram(pd)
        pd.actions.forEach { n, a -> pd.scope.registerAction(a) }
        return null
    }

    override fun visit(fbd: FunctionBlockDeclaration): Any? {
        goIntoScope(fbd)
        scope.registerFunctionBlock(fbd)
        fbd.actions.forEach { scope.registerAction(it)  }
        fbd.methods.forEach { scope.registerMethod(it); }
        goOutoScope()
        return null
    }

    override fun visit(functionDeclaration: FunctionDeclaration): Any? {
        goIntoScope(functionDeclaration)
        scope.registerFunction(functionDeclaration)
        goOutoScope()
        return null
    }

    override fun visit(subRangeTypeDeclaration: SubRangeTypeDeclaration): Any? {
        scope.registerType(subRangeTypeDeclaration)
        return null
    }

    override fun visit(variableDeclaration: VariableDeclaration): Any? {
        //weigl: do not register anonymous datatype, or references to data types.
        /*
        if (variableDeclaration.getTypeDeclaration() instanceof ArrayTypeDeclaration)
            variableDeclaration.getTypeDeclaration().accept(this);
        return super.visit(variableDeclaration);
        */
        return null
    }

    override fun visit(simpleTypeDeclaration: SimpleTypeDeclaration): Any? {
        scope.registerType(simpleTypeDeclaration)
        return null //super.visit(simpleTypeDeclaration)
    }

    override fun visit(ptd: PointerTypeDeclaration): Any? {
        scope.registerType(ptd)
        return super.visit(ptd)
    }

    override fun visit(clazz: ClassDeclaration): Any? {
        scope.registerClass(clazz)
        goIntoScope(clazz)
        clazz.methods.forEach { scope.registerMethod(it) }
        goOutoScope()
        return super.visit(clazz)
    }

    override fun visit(interfaceDeclaration: InterfaceDeclaration): Any? {
        //goIntoScope(interfaceDeclaration)
        scope.registerInterface(interfaceDeclaration)
        return null //super.visit(interfaceDeclaration)
    }

    override fun visit(arrayTypeDeclaration: ArrayTypeDeclaration): Any? {
        scope.registerType(arrayTypeDeclaration)
        return null
    }

    /**
     * {@inheritDoc}
     */
    override fun visit(enumerationTypeDeclaration: EnumerationTypeDeclaration): Any? {
        scope.registerType(enumerationTypeDeclaration)
        return null
    }

    /**
     * {@inheritDoc}
     */
    override fun visit(stringTypeDeclaration: StringTypeDeclaration): Any? {
        scope.registerType(stringTypeDeclaration)
        return null
    }

    override fun visit(typeDeclarations: TypeDeclarations): Any? {
        for (td in typeDeclarations)
            td.accept(this)
        return null
    }

    override fun visit(structureTypeDeclaration: StructureTypeDeclaration): Any? {
        scope.registerType(structureTypeDeclaration)
        return null
    }

    override fun visit(gvl: GlobalVariableListDeclaration): Any? {
        gvl.scope = scope // global variables does not open an own scope.
        scope.addVariables(gvl.scope)
        //return super.visit(gvl)
        return null
    }

    private fun goIntoScope(hasScope: HasScope) {
        hasScope.scope.parent = scope
        scope = hasScope.scope
    }

    private fun goOutoScope() {
        scope = if (scope.parent != null) scope.parent!! else globalScope
    }
}
